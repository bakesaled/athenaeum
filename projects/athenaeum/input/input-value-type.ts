export type AthInputValueType = 'number' | 'text' | null;
