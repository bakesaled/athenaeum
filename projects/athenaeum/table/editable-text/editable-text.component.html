<ng-container *ngIf="(editMode$ | async) === true">
  <form [formGroup]="inputFormGroup" (ngSubmit)="onSubmit()">
    <mat-form-field appearance="outline" [ngStyle]="{ width: getWidth() }">
      <input matInput formControlName="input" athFocusable (blur)="onBlur()" />
    </mat-form-field>
  </form>
</ng-container>
<ng-container *ngIf="(editMode$ | async) === false">
  {{ value }}
</ng-container>
