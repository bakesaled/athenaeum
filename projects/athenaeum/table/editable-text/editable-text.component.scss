:host {
  display: block;

  .mat-form-field {
    line-height: unset;
  }
}

:host::ng-deep .mat-form-field-wrapper {
  padding-bottom: 0;
}

:host::ng-deep .mat-form-field-appearance-outline .mat-form-field-wrapper {
  margin: 0;
}

:host::ng-deep .mat-form-field-appearance-outline .mat-form-field-flex {
  padding-right: 1px;
  padding-left: 1px;
}

:host::ng-deep .mat-form-field-infix {
  border-top: 0;
}

:host::ng-deep .mat-form-field-appearance-outline .mat-form-field-outline-end {
  border-radius: 0 !important;
  border-style: none;
}
:host::ng-deep
  .mat-form-field-appearance-outline
  .mat-form-field-outline-start {
  border-radius: 0 !important;
  border-style: none;
}
:host::ng-deep .mat-form-field-appearance-outline .mat-form-field-outline-gap {
  border-radius: 0 !important;
  border-bottom: none;
}
