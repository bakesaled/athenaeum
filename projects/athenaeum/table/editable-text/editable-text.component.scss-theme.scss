@import '~@angular/material/theming';

@mixin editable-text-component-color($config-or-theme) {
  $config: mat-get-color-config($config-or-theme);
  //$primary: map-get($config, primary);
  $foreground: map-get($config, foreground);

  .mat-form-field {
    border-left: mat-color($foreground, divider) solid 1px;
    border-right: mat-color($foreground, divider) solid 1px;
  }
}

@mixin editable-text-component-theme($theme) {
  $color: mat-get-color-config($theme);

  @if $color != null {
    @include editable-text-component-color($color);
  }
}
