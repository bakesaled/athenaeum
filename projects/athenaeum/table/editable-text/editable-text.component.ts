import {
  Component,
  OnInit,
  ChangeDetectionStrategy,
  Input,
  ElementRef,
  OnDestroy,
  AfterViewInit,
  ChangeDetectorRef,
  Output,
  EventEmitter,
} from '@angular/core';
import { BehaviorSubject, fromEvent, Subject } from 'rxjs';
import { switchMapTo, takeUntil } from 'rxjs/operators';
import { FormControl, FormGroup } from '@angular/forms';

@Component({
  selector: 'ath-editable-text',
  templateUrl: './editable-text.component.html',
  styleUrls: ['./editable-text.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
  host: {
    class: 'ath-editable-text',
  },
})
export class AthEditableTextComponent
  implements OnInit, AfterViewInit, OnDestroy {
  private destroySubject = new Subject();
  private hostElementWidth;
  private inputVal: string;

  editModeSubject = new BehaviorSubject<boolean>(false);

  editMode$ = this.editModeSubject.asObservable();

  @Output() update: EventEmitter<any> = new EventEmitter<any>();

  @Input()
  get value(): string {
    return this.inputVal;
  }
  set value(val: string) {
    this.inputVal = val;
    this.inputFormGroup.controls.input.setValue(val);
    this.changeDetector.detectChanges();
  }
  inputFormGroup: FormGroup;

  constructor(
    private hostEl: ElementRef,
    private changeDetector: ChangeDetectorRef
  ) {
    this.inputFormGroup = new FormGroup({
      input: new FormControl(),
    });
  }

  ngOnInit(): void {
    this.editModeHandler();
  }

  ngAfterViewInit(): void {
    setTimeout(() => {
      this.hostElementWidth = this.hostEl.nativeElement.offsetWidth;
    });
  }

  ngOnDestroy(): void {
    this.destroySubject.next();
  }

  private editModeHandler(): void {
    const clickOutside$ = fromEvent(document, 'click').pipe();

    this.editMode$
      .pipe(switchMapTo(clickOutside$), takeUntil(this.destroySubject))
      .subscribe((val) => {
        if (
          this.isOrContainsClickTarget(
            this.hostEl.nativeElement,
            val.target as HTMLElement
          )
        ) {
          this.toEditMode();
        }
      });
  }

  toEditMode(): void {
    if (!this.editModeSubject.getValue()) {
      this.editModeSubject.next(true);
    }
  }

  toViewMode(): void {
    // Only fire if in edit mode
    if (this.editModeSubject.getValue()) {
      // Only fire update if value has changed
      if (this.inputVal !== this.inputFormGroup.value.input) {
        this.inputVal = this.inputFormGroup.value.input;
        this.update.emit(this.inputFormGroup.value.input);
      }

      this.editModeSubject.next(false);
    }
  }

  getWidth(): string {
    this.hostEl.nativeElement.parentElement.style.width = `${this.hostElementWidth}px`;

    // Subtract 2 for border.
    return `${this.hostElementWidth - 2}px`;
  }

  onBlur(): void {
    this.toViewMode();
  }

  onSubmit(): void {
    this.toViewMode();
  }

  private isOrContainsClickTarget(
    element: HTMLElement,
    clickTarget: HTMLElement
  ): boolean {
    return element === clickTarget || element.contains(clickTarget);
  }
}
