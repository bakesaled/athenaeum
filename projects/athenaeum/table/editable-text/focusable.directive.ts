import { AfterViewInit, Directive, ElementRef } from '@angular/core';

@Directive({
  selector: '[athFocusable]',
})
export class FocusableDirective implements AfterViewInit {
  constructor(private hostEl: ElementRef) {}

  ngAfterViewInit(): void {
    this.hostEl.nativeElement.focus();
    this.hostEl.nativeElement.select();
  }
}
