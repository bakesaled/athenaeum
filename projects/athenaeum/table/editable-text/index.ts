export * from './editable-text.component';
export * from './focusable.directive';
