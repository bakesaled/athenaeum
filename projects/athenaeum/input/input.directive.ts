import {
  Directive,
  ElementRef,
  forwardRef,
  HostListener,
  Input,
} from '@angular/core';
import { MAT_INPUT_VALUE_ACCESSOR } from '@angular/material/input';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { formatNumber } from '@angular/common';
import { AthInputValueType } from './input-value-type';

/**
 * A directive that adds helpful behavior to input and textarea elements.
 */
@Directive({
  selector: 'input[athInput], textarea[athInput]',
  exportAs: 'athInput',
  providers: [
    { provide: MAT_INPUT_VALUE_ACCESSOR, useExisting: AthInputDirective },
    {
      provide: NG_VALUE_ACCESSOR,
      useExisting: forwardRef(() => AthInputDirective),
      multi: true,
    },
  ],
  host: {
    class: 'ath-input-element',
    '[style.text-align]': '"right"',
  },
})
export class AthInputDirective {
  private val: string;

  /**
   * Value assigned to the element (MatFormField).
   */
  @Input()
  get value(): string {
    return this.val;
  }

  set value(value: string | null) {
    this.val = value;
    this.formatValue(value);
  }

  locale = 'en-US';

  /**
   * Decimal representation options, specified by a string in the following format:
   * `{minIntegerDigits}.{minFractionDigits}-{maxFractionDigits}`.
   */
  @Input()
  digitInfo: string;

  /**
   * Type of value: 'number' or 'text' (default: 'text');
   */
  @Input()
  valueType: AthInputValueType;

  constructor(private elementRef: ElementRef<HTMLInputElement>) {}

  @HostListener('blur')
  onBlur(): void {
    if (this.val) {
      this.formatValue(this.val);
    }
  }

  @HostListener('focus')
  onFocus(): void {
    this.unformatValue();
  }

  @HostListener('input', ['$event.target.value'])
  onInput(value): void {
    if (this.valueType === 'number') {
      this.val = this.unformatNumber(value);
    } else {
      this.val = value;
    }
    this.onChange(this.val);
  }

  writeValue(value: any): void {
    this.val = value;
    if (value) {
      this.formatValue(this.val);
    }
  }

  onChange(_): void {}

  registerOnChange(fn: (value: any) => void): void {
    this.onChange = fn;
  }

  registerOnTouched(): void {}

  unformatNumber(target: string): string {
    if (target) {
      return target.replace(/[^\d.-]/g, '');
    } else {
      return '';
    }
  }

  private formatValue(value: any): void {
    if (this.valueType === 'number') {
      this.elementRef.nativeElement.value = formatNumber(
        value,
        this.locale,
        this.digitInfo
      );
    }
  }

  private unformatValue(): void {
    if (this.valueType === 'number') {
      this.elementRef.nativeElement.value = this.unformatNumber(
        this.elementRef.nativeElement.value
      );
    }
  }
}
