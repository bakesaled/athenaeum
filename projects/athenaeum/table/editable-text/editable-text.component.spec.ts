import { ComponentFixture, TestBed } from '@angular/core/testing';

import { AthEditableTextComponent } from './editable-text.component';
import { MatFormFieldModule } from '@angular/material/form-field';
import { MatInputModule } from '@angular/material/input';
import { ReactiveFormsModule } from '@angular/forms';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';

describe('EditableTextComponent', () => {
  let component: AthEditableTextComponent;
  let fixture: ComponentFixture<AthEditableTextComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [AthEditableTextComponent],
      imports: [
        MatFormFieldModule,
        MatInputModule,
        ReactiveFormsModule,
        NoopAnimationsModule,
      ],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(AthEditableTextComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should render', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should render in view mode', () => {
    const textEl = fixture.elementRef.nativeElement.querySelector('form');
    expect(textEl).toBeNull();
  });

  it('should render in edit mode', () => {
    component.editModeSubject.next(true);
    fixture.detectChanges();
    const textEl = fixture.elementRef.nativeElement.querySelector('form');
    expect(textEl).not.toBeNull();
  });
});
