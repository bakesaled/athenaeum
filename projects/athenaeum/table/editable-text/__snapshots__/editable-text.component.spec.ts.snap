// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableTextComponent should render 1`] = `
<ath-editable-text
  changeDetector={[Function ViewRef_]}
  destroySubject={[Function Subject]}
  editMode$={[Function Observable]}
  editModeSubject={[Function BehaviorSubject]}
  hostEl={[Function ElementRef]}
  inputFormGroup={[Function FormGroup]}
  update={[Function EventEmitter_]}
>
   
  
</ath-editable-text>
`;
